ui/app/pages/permissions-connect/permissions-connect.component.js

```diff
@@ -53,43 +49,75 @@ export default class PermissionConnect extends Component {
     selectedAccountAddresses: this.props.accounts.length === 1
       ? new Set([this.props.accounts[0].address])
       : new Set(),
-    permissionAccepted: null,
+    permissionsApproved: null,
     origin: this.props.origin,
+    targetDomainMetadata: this.props.targetDomainMetadata || {},
   }
 
   beforeUnload = () => {
     const { permissionsRequestId, rejectPermissionsRequest } = this.props
-    const { permissionAccepted } = this.state
+    const { permissionsApproved } = this.state
 
-    if (permissionAccepted === null && permissionsRequestId) {
+    if (permissionsApproved === null && permissionsRequestId) {
       rejectPermissionsRequest(permissionsRequestId)
     }
   }
 
   removeBeforeUnload = () => {
     const environmentType = getEnvironmentType()
-    if (
-      environmentType === ENVIRONMENT_TYPE_FULLSCREEN ||
-      environmentType === ENVIRONMENT_TYPE_NOTIFICATION
-    ) {
+    if (environmentType === ENVIRONMENT_TYPE_NOTIFICATION) {
       window.removeEventListener('beforeunload', this.beforeUnload)
     }
   }
 
+  componentDidMount () {
+    const {
+      getCurrentWindowTab,
+      getRequestAccountTabIds,
+      permissionsRequest,
+      history,
+    } = this.props
+    getCurrentWindowTab()
+    getRequestAccountTabIds()
+
+    if (!permissionsRequest) {
+      return history.push(DEFAULT_ROUTE)
+    }
+
+    const environmentType = getEnvironmentType()
+    if (environmentType === ENVIRONMENT_TYPE_NOTIFICATION) {
+      window.addEventListener('beforeunload', this.beforeUnload)
+    }
+  }
+
   componentDidUpdate (prevProps) {
-    const { permissionsRequest, lastConnectedInfo } = this.props
-    const { redirecting, origin } = this.state
+    const {
+      permissionsRequest,
+      lastConnectedInfo,
+      targetDomainMetadata,
+    } = this.props
+    const {
+      redirecting,
+      origin,
+      targetDomainMetadata: savedMetadata,
+    } = this.state
+
+    if (
+      permissionsRequest &&
+      savedMetadata.name !== targetDomainMetadata?.name
```

Comment:
Keeping state in sync with props is better handled by `getDeriveStateFromProps`. `componentDidUpdate` runs _after_ the update, so this `targetDomainMetadata` state will be delayed in updating.

Reply:
d8e7fb4